### PR TITLE
Fast Copter Fences 

### PR DESCRIPTION
**Pull request by andyp1per**
An attempt at checking fences quickly to avoid massive breaches

Adds support for auto-enabling fence floors on copter

Current master is fixed at 3Hz so if you are going 30m/s you will be 10m before the fence before it is even checked

Demo: https://youtu.be/8umjR5EAKfg

### CODE REVIEW

**Comment by andyp1per** [comment]
I am testing using this script: https://gist.github.com/andyp1per/85e5e29dac031e4c93d6e83d8f71641f

Here is an example log: [00000056.zip](https://github.com/ArduPilot/ardupilot/files/13549271/00000056.zip)

The copter flies through the fence at a very high speed and is about 50m past it before the fence kicks in. On the way down the fence just about kicks in before it hits the ground.

File: libraries/AC_Fence/AC_Fence.cpp [hunk 1]
```diff
@@ -470,6 +471,33 @@ bool AC_Fence::check_fence_alt_min()
...
+    if (!(_enabled_fences & AC_FENCE_TYPE_ALT_MIN)) {
+        // not enabled
+        return false;
+    }
+
+    AP::ahrs().get_relative_position_D_home(_curr_alt);
+    _curr_alt = -_curr_alt; // translate Down to Up
+
+    // check if we are under the altitude fence
+    if (_curr_alt >= _alt_min) {
```
Comment (tridge) [inline libraries/AC_Fence/AC_Fence.cpp:492]: needs a margin

Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp]: Fixed

File: ArduCopter/mode.cpp [hunk 2]
```diff
@@ -368,10 +368,13 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
 #endif
 
 #if AP_FENCE_ENABLED
-    // pilot requested flight mode change during a fence breach indicates pilot is attempting to manually recover
-    // this flight mode change could be automatic (i.e. fence, battery, GPS or GCS failsafe)
-    // but it should be harmless to disable the fence temporarily in these situations as well
-    fence.manual_recovery_start();
+    // scripted changes are by definition not manual
+    if (reason != ModeReason::SCRIPTING) {
```
Comment (peterbarker) [inline ArduCopter/mode.cpp:372]: If fence action is zero we really ought to do nothing - including this manual recovery work.

Comment (andyp1per) [reply ArduCopter/mode.cpp]: Fixed

**Comment by Hwurzburg** [comment]
does this fix plane also? I think mine fixes both

**Comment by andyp1per** [comment]
> does this fix plane also? I think mine fixes both....note also that autoenable handles floor differently for each case

Well it fixes the bugs I want fixed of which there are a few. Copter is actually missing the autoenable feature which this implements amongst other things. What is the bug you think that needs fixing in plane?

**Comment by Hwurzburg** [comment]
Fences have the following issues:
- MIN_ALT fence records breach even if floor is not enabled
- MIN_ALT floor always disabled on boot irrespective FENCE_ENABLE, never gets enabled unless: autoenabled, or overt enable(GCS or switch)
- MIN_ALT floor stays disabled after disable floor for landing, used by NAV_LAND, QLAND and other VTOL landings in Plane, LAND, Copter RTL
- Plane automatically ignores any breach while performing a breach action. If landing occurs, MIN_ALT breach is recorded but  not cleared after landing, preventing mode changes if mode change prevention is enabled (Copter allows mode changes always). This prevents mode changes after landing.
- AutoEnable on arming or not using AutoEnabled never disables a floor for landing. So all the above situations which disable floor, wont.
Any changes to  AC_Fence module without considering how it may impact Plane or future fixes for plane fence issues is a problem IMHO

Having floor always enable once MIN_ALT is reached by itself wont impact future fixes and only impacts the MIN_ALT fence, whether that fence is setup or not...but it must be able to do this after disables from any altitude, and after landing for both Copter and Plane

File: ArduCopter/fence.cpp [hunk 3]
```diff
@@ -24,7 +24,8 @@ void Copter::fence_check()
     if (new_breaches) {
 
         if (!copter.ap.land_complete) {
```
Comment (IamPete1) [inline ArduCopter/fence.cpp:26]: Its a bit odd that we don't print if landed. Its the current behavior of course. But now it will print when the breach clears in all caces, so you can end up reporting the clear but not the breach.

File: ArduCopter/mode.cpp [hunk 4]
```diff
@@ -368,10 +368,12 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
 #endif
 
 #if AP_FENCE_ENABLED
-    // pilot requested flight mode change during a fence breach indicates pilot is attempting to manually recover
-    // this flight mode change could be automatic (i.e. fence, battery, GPS or GCS failsafe)
-    // but it should be harmless to disable the fence temporarily in these situations as well
-    fence.manual_recovery_start();
+    if (fence.get_action() != AC_FENCE_ACTION_REPORT_ONLY ) {
```
Comment (IamPete1) [inline ArduCopter/mode.cpp:371]:
```suggestion
    if (fence.get_action() != AC_FENCE_ACTION_REPORT_ONLY) {
```

File: libraries/AC_Fence/AC_Fence.cpp [hunk 5]
```diff
@@ -470,6 +471,33 @@ bool AC_Fence::check_fence_alt_min()
...
+        // not enabled
+        return false;
+    }
+
+    AP::ahrs().get_relative_position_D_home(_curr_alt);
+    _curr_alt = -_curr_alt; // translate Down to Up
+
+    // check if we are under the altitude fence
+    if (_curr_alt >= _alt_min + _margin) {
+        enable_floor();
```
Comment (IamPete1) [inline libraries/AC_Fence/AC_Fence.cpp:493]: Should we also do a send text?

File: libraries/AC_Fence/AC_Fence.cpp [hunk 6]
```diff
@@ -126,19 +126,20 @@ const AP_Param::GroupInfo AC_Fence::var_info[] = {
...
     // @DisplayName: Fence Auto-Enable
     // @Description: Auto-enable of fence
-    // @Values: 0:AutoEnableOff,1:AutoEnableOnTakeoff,2:AutoEnableDisableFloorOnLanding,3:AutoEnableOnlyWhenArmed
+    // @Values{Plane}: 0:AutoEnableOff,1:AutoEnableOnTakeoff,2:AutoEnableDisableFloorOnLanding,3:AutoEnableOnlyWhenArmed
+    // @Values{Copter}: 0:AutoEnableOff,1:AutoEnableFloorOnTakeoff
     // @Range: 0 3
     // @Increment: 1
     // @User: Standard
-    AP_GROUPINFO_FRAME("AUTOENABLE", 10, AC_Fence, _auto_enabled, static_cast<uint8_t>(AutoEnable::ALWAYS_DISABLED), AP_PARAM_FRAME_PLANE),
+    AP_GROUPINFO_FRAME("AUTOENABLE", 10, AC_Fence, _auto_enabled, static_cast<uint8_t>(AutoEnable::ALWAYS_DISABLED), AP_PARAM_FRAME_PLANE | AP_PARAM_FRAME_COPTER),
```
Comment (IamPete1) [inline libraries/AC_Fence/AC_Fence.cpp:137]:
```suggestion
    AP_GROUPINFO_FRAME("AUTOENABLE", 10, AC_Fence, _auto_enabled, static_cast<uint8_t>(AutoEnable::ALWAYS_DISABLED), AP_PARAM_FRAME_PLANE | AP_PARAM_FRAME_COPTER | AP_PARAM_FRAME_TRICOPTER | AP_PARAM_FRAME_HELI),
```

File: libraries/AC_Fence/AC_Fence.cpp [hunk 7]
```diff
@@ -126,19 +126,20 @@ const AP_Param::GroupInfo AC_Fence::var_info[] = {
...
 
     // @Param{Plane}: OPTIONS
     // @DisplayName: Fence options
-    // @Description: 0:Disable mode change following fence action until fence breach is cleared. When bit 1 is set the allowable flight areas is the union of all polygon and circle fence areas instead of the intersection, which means a fence breach occurs only if you are outside all of the fence areas.
-    // @Bitmask: 0:Disable mode change following fence action until fence breach is cleared, 1:Allow union of inclusion areas
+    // @Description: 0:Disable mode change following fence action until fence breach is cleared. When bit 1 is set the allowable flight areas is the union of all polygon and circle fence areas instead of the intersection, which means a fence breach occurs only if you are outside all of the fence areas. When bit 2 is set any active minimum altitude floor will be disabled.
+    // @Bitmask: 0:Disable mode change following fence action until fence breach is cleared, 1:Allow union of inclusion areas, 2:Auto-disable floor fence on disarming
```
Comment (IamPete1) [inline libraries/AC_Fence/AC_Fence.cpp:142]: Why would you have this bit ever disabled? I would have thought you would also want the floor disabled when disarming if you have `AutoEnable::ALWAYS_ENABLED`

Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp:142]: @IamPete1 because the tests check for the current disarmed behaviour. We can change the tests of course, but we need to be all agreed that this is the correct behaviour - because it is a change

File: libraries/AC_Fence/AC_Fence.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp:137]: Thanks!

File: libraries/AC_Fence/AC_Fence.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp:493]: Yeah maybe - I was allowing most of these to be reported in the vehicle code rather than in the fences library

File: ArduCopter/mode.cpp (hunk 4 above)
Comment (andyp1per) [reply ArduCopter/mode.cpp:371]: Thanks!

File: libraries/AC_Fence/AC_Fence.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp:493]: Fixed

**Comment by andyp1per** [comment]
@Hwurzburg 

> * MIN_ALT fence records breach even if floor is not enabled

I think this is fixed by this PR, but its possible there is something special going on in plane.

> * MIN_ALT floor always disabled on boot irrespective FENCE_ENABLE, never gets enabled unless: autoenabled, or overt enable(GCS or switch)

Yes, this is the main fix of this PR. Without this MIN_ALT is mostly useless on Copter.

> * MIN_ALT floor stays disabled after disable floor for landing, used by NAV_LAND, QLAND and other VTOL landings in Plane, LAND, Copter RTL

Can you expand on this? Current behaviour is that fence remains enable but is ignored for landing so that you cannot rearm. This PR provides an option which automatically clears the breach once you have landed so that you can rearm without rebooting.
 
> * Plane automatically ignores any breach while performing a breach action. If landing occurs, MIN_ALT breach is recorded but  not cleared after landing, preventing mode changes if mode change prevention is enabled (Copter allows mode changes always). This prevents mode changes after landing.

For copter see above - I think this is partially fixed (second part). I'm not sure that the first part is the wrong behavior - once your breach action has engaged you want it to stay engaged. I did fix a bug which preventing scripting seeing new breaches.

> * AutoEnable on arming or not using AutoEnabled never disables a floor for landing. So all the above situations which think they disable floor, wont.

I don't quite get this one but I think its plane specific as auto enable is not currently supported on copter without this PR. With this PR I am able to land ok (at least in RTL).

**Comment by Hwurzburg** [comment]
@andyp1per any change to AC_Fences must be THOROUGHLY tested for impacts on Plane....it took me two days of SITL sims on just MIN_ALT fence to find all the hidden anomalies...

**Comment by andyp1per** [comment]
> @andyp1per any change to AC_Fences must be THOROUGHLY tested for impacts on Plane....it took me two days of SITL sims on just MIN_ALT fence to find all the hidden anomalies...

Sure. The solution here, of course, is making sure that the autotests cover the edge cases. I have added some tests but more can be added - be good to know what tests you think are missing. I also have spent a long time flying this in RF, that also enabled me to uncover many of these hidden "gems". Clearly the fencing code is not as widely used as we suspect otherwise they would have been revealed sooner.

File: ArduCopter/Copter.cpp [hunk 8]
```diff
@@ -147,6 +147,9 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
 
     SCHED_TASK(rc_loop,              250,    130,  3),
     SCHED_TASK(throttle_loop,         50,     75,  6),
+#if AP_FENCE_ENABLED
+    SCHED_TASK(fence_check,           25,    100,  7),
```
Comment (tridge) [inline ArduCopter/Copter.cpp:151]: might be worth moving fence checking to an IO callback, or add a FENCE_OPTIONS bit for fast fence checking

File: libraries/AC_Fence/AC_Fence.cpp [hunk 9]
```diff
@@ -470,6 +471,34 @@ bool AC_Fence::check_fence_alt_min()
...
+    // altitude fence check
+    if (!(_enabled_fences & AC_FENCE_TYPE_ALT_MIN)) {
+        // not enabled
+        return false;
+    }
+
+    AP::ahrs().get_relative_position_D_home(_curr_alt);
+    _curr_alt = -_curr_alt; // translate Down to Up
+
+    // check if we are under the altitude fence
```
Comment (tridge) [inline libraries/AC_Fence/AC_Fence.cpp:491]: over not under?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by rmackay9** [comment]
We discussed at length on the dev call but disagreed about how fast the fence check could be run.  Various suggestions came up re adding a compile time update rate parameter, a parameter option, dependent upon the CPU time etc.

I think it would be good to first get a baseline of how expensive the fence check is.  For example, how complex can the fence be (e.g. how many polygon points) and still complete within 1ms on an F4 CPU?

File: libraries/AC_Fence/AC_Fence.cpp [hunk 10]
```diff
@@ -470,6 +471,34 @@ bool AC_Fence::check_fence_alt_min()
     return false;
 }
 
+
+/// auto enable fence floor
+bool AC_Fence::auto_enable_fence_floor()
+{
+    if (_floor_enabled) {
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp]: Moved to https://github.com/ArduPilot/ardupilot/pull/25875

File: libraries/AC_Fence/AC_Fence.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AC_Fence/AC_Fence.cpp]: Fixed thanks

**Comment by andyp1per** [comment]
Fence fixes split out into https://github.com/ArduPilot/ardupilot/pull/25875

**Comment by andyp1per** [comment]
On a CubeBlack I get about 10us per cycle for general fence checks, if I add a poly fence with 25 points It goes to about 22us percycle, if I increase this to include an exclusion fence with another 42 points I get about 33us. So being ultra conservative lets say we can get 1 point per us that means we can have a 1000 point fence for 1ms on an F4, but in all likelihood considerably more than this.

**Review by rmackay9 — APPROVED** [review]
Thanks for the testing.  Personally I'd make it 10hz because of the law of diminishing returns but it's probably fine at 25hz too.  A very fast copter flying at 20m/s will get 6.6m breach currently, only 2m at 10hz, 0.8m at 25hz.

**Review by tridge — APPROVED** [review]
